src/upnp.cpp

```diff
@@ -1566,6 +1569,8 @@ void upnp::on_expire(error_code const& ec)
 	COMPLETE_ASYNC("upnp::on_expire");
 	if (ec) return;
 
+	std::shared_ptr<upnp> me(self());
```

Comment:
I don't believe there's anything in this function that may cause the last reference to this object to be released. do you?